### DISCUSSION TITLE
main menu

### DISCUSSION THREAD

**Discussion by DexterTaha** [pybricks/support#1401] [General]
hi and you hope you are doing well, I am participating in the FLL, I use  a robot inventor hub and  I have 11 runs and code and waste time changing between code btw I use a main menu but it take time so there is any method to when a code finished passe automatically to the next code after clicking on a button.
This the my main code:
```python
from pybricks.tools import hub_menu

# Choose a letter.
selected = hub_menu("1", "2", "P", "O", "4", "5", "6", "7", "8", "9", "R")

# Based on the selection, run a program.
if selected == "R":
    import R

elif selected == "1":
    import R1

elif selected == "2":
    import R2

elif selected == "4":
    import R4

elif selected == "P":
    import P

elif selected == "O":
    import O

elif selected == "5":
    import R5

elif selected == "6":
    import R6

elif selected == "7":
    import R7

elif selected == "8":
    import R8

elif selected == "9":
    import R9
```
for exmple when i finish the run 7 it take a lot of time to run code 8

**Comment by laurensvalk** [comment]
Hi! Here are some suggestions to make more advanced menus and shared functions for more flexibility:
https://github.com/orgs/pybricks/discussions/1387

**Comment by bcfletcher** [comment]
This is the approach I used with my FLL team for Submerged

```
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch, multitask, run_task, hub_menu
from umath import sin, pi
from SM_robot_core_rev10 import stopEverything
from SM_t0_rev9 import T0_Run
from SM_t1_rev15 import T1_Run
from SM_t2_rev19 import T2_Run
from SM_t3_rev8 import T3_Run
from SM_t4_rev10 import T4_Run
from SM_t5_rev25 import T5_Run


print("OMMS South FLL Team 2 Submerged Menu")
print("    Setting Up...")
# Initialize the hub.
hub = PrimeHub()

hub.system.set_stop_button(Button.CENTER)

# Make a menu to choose a letter. You can also use numbers.

# Based on the selection, run a program.
print("    Running Menu")
hub.light.on(Color.BLUE)

while True:
    selected = hub_menu("0","1","2","3","4","5","X")

    try:
        if selected == "0":
            run_task(T0_Run())
            run_task(stopEverything())
        elif selected == "1":
            run_task(T1_Run())
            run_task(stopEverything())
        elif selected == "2":
            run_task(T2_Run())  
            run_task(stopEverything())
        elif selected == "3":
            run_task(T3_Run())  
            run_task(stopEverything())
        elif selected == "4":
            run_task(T4_Run())  
            run_task(stopEverything())
        elif selected == "5":
            run_task(T5_Run())  
            run_task(stopEverything())
        elif selected == "X":
            run_task(stopEverything())
            break

    except BaseException as menuException:
        print("Stop was Pressed or a Critical Error Occured. Stopping all motors.")
        print(menuException)
        run_task(stopEverything())
        break

print ("End of Menu - if you see this, there is a problem")

async def stopEverything():
    print("Stop All Motors")
    driveBase.use_gyro(False)
    leftMotor.stop()
    wait(250)
    rightMotor.stop()
    wait(250)
    topAttachment.stop()
    wait(250)
    frontAttachment.stop()
    wait(250)
    print("All motors should be stopped")
```

Each SM_t#_rev##.py file has a T#_Run() which replaces the traditional main() function.

at the end of each .py file we have a run_task(T#_Run()) command commented out.  we uncomment the run_task() when we want to run the task manually on the robot.

previously we used hub.system.shutdown() in the except along with the other items  but it ended up wiping all the programs from the reboot for some unknown reason.

**Reply by deopark** [reply]
Is there a way to force it to stop while the partial mission is running? (not shutdown)

**Reply by Unknown8me** [reply]
Ours didn't work until we added an Axis to the pybricks.parameters just letting anyone know if you having trouble.

**Reply by bcfletcher** [reply]
yes, press the Big Round Button. it will stop of the robot.

**Comment by Digital-Dominators** [comment]
Here is our way to implement it.

We use the hub.system.storage function to store the previous executed route to the storage which is retained between routes (and even you power off the hub).

The byte stored in the storage is simply a number (e.g. 1-11). And at the beginning of your menu file, read the number in the storage.  Based on the number you read (representing the last executed route), you set different menus. e.g.

```
if previous_route == 1:
   selected = hub_menu("2", "3", "4", "5", "1")
elif previous_route == 1:
   selected = hub_menu("3", "4", "5", "1", "2")
......

```

Then in each selection, besides running your code, add the previous route by 1 and refresh the hub.system.storage with the new number. 

In this way, after you finish your run 1 and press the start button again,  the menu will automatically be set to "2".  You don't need to use the left/right button to change selection.  Meanwhile, you still keep the possibility to switch back to previous routes using the right button if you want to re-run it.

**Reply by dlech** [reply]
Nice solution. :+1: 

Nothing wrong with the way you did it, but for those of us who like little puzzles, it could be interesting to create an iterator to generate the args instead of making an if statement to avoid having to type out all of the combinations.

Untested:

```python

NUM_ROUTES = 5

def rotate_items(prev, last):
    # emit items starting with the one after prev up to last
    for i in range(prev + 1, last + 1):
        yield str(i)
    # then wrap around and emit items 1 to prev
    for i in range(1, prev + 1):
        yield str(i)

...

# start menu with "1"
previous_route = NUM_ROUTES

while True:
    selected = hub_menu(*rotate_items(previous_route, NUM_ROUTES))
    ...
```

**Reply by Digital-Dominators** [reply]
Thanks for the suggestion!
